bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java

```diff
@@ -0,0 +1,322 @@
+
+        long latencyNanos = MathUtils.elapsedNanos(requestTimeNanos);
+        if (code != BKException.Code.OK) {
+            LOG.error(
+                    "Read of ledger entry failed: L{} E{}-E{}, Sent to {}, "
+                            + "Heard from {} : bitset = {}, Error = '{}'. First unread entry is ({}, rc = {})",
+                    lh.getId(), startEntryId, endEntryId, sentToHosts, heardFromHosts, heardFromHostsBitSet,
+                    BKException.getMessage(code), startEntryId, code);
+            clientCtx.getClientStats().getReadOpLogger().registerFailedEvent(latencyNanos, TimeUnit.NANOSECONDS);
+            // release the entries
+
+            request.close();
+            future.completeExceptionally(BKException.create(code));
+        } else {
+            clientCtx.getClientStats().getReadOpLogger().registerSuccessfulEvent(latencyNanos, TimeUnit.NANOSECONDS);
+            future.complete(LedgerEntriesImpl.create(request.entries));
+        }
+    }
+
+    BatchedReadOp parallelRead(boolean enabled) {
+        this.parallelRead = enabled;
+        return this;
+    }
+
+    @Override
+    public void readEntriesComplete(int rc, long ledgerId, long startEntryId, ByteBufList bufList, Object ctx) {
+        final ReadContext rctx = (ReadContext) ctx;
+        final BatchedLedgerEntryRequest entry = (BatchedLedgerEntryRequest) rctx.entry;
+
+        if (rc != BKException.Code.OK) {
+            entry.logErrorAndReattemptRead(rctx.bookieIndex, rctx.to, "Error: " + BKException.getMessage(rc), rc);
+            return;
+        }
+
+        heardFromHosts.add(rctx.to);
+        heardFromHostsBitSet.set(rctx.bookieIndex, true);
+
+        bufList.retain();
+        // if entry has completed don't handle twice
+        if (entry.complete(rctx.bookieIndex, rctx.to, bufList)) {
+            if (!isRecoveryRead) {
+                // do not advance LastAddConfirmed for recovery reads
+                lh.updateLastConfirmed(rctx.getLastAddConfirmed(), 0L);
+            }
+            submitCallback(BKException.Code.OK);
+        } else {
+            bufList.release();
+        }
+    }
+
+    void sendReadTo(int bookieIndex, BookieId to, BatchedLedgerEntryRequest entry) throws InterruptedException {
+        if (lh.throttler != null) {
+            lh.throttler.acquire();
+        }
+        if (isRecoveryRead) {
+            int flags = BookieProtocol.FLAG_HIGH_PRIORITY | BookieProtocol.FLAG_DO_FENCING;
+            clientCtx.getBookieClient().readEntries(to, lh.ledgerId, entry.eId,
+                    maxCount, maxSize, this, new ReadContext(bookieIndex, to, entry), flags, lh.ledgerKey);
+        } else {
+            clientCtx.getBookieClient().readEntries(to, lh.ledgerId, entry.eId, maxCount, maxSize,
+                    this, new ReadContext(bookieIndex, to, entry), BookieProtocol.FLAG_NONE);
+        }
+    }
+
+    abstract class BatchedLedgerEntryRequest extends LedgerEntryRequest {
+
+        final long lId;
```

Comment:
Do we need to store the ledgerId?

Reply:
make sense. Please add a comment on the `lId` to avoid being removed in the future.